### PR TITLE
Sourcebuild perf perhaps

### PR DESCRIPTION
**Pull request by KevinRansom**
Fixes: https://github.com/dotnet/fsharp/issues/16833
Fixes: https://github.com/dotnet/fsharp/issues/16969

- Speeds up sourcebuild by 10-20 % which is pretty decent.
- The effect on repo builds is mostly neutral, perhaps marginally faster builds.
- Switches proto compiler to a Ready2Run compiled coreclr compiler
- Fixes the deterministic build test to work correctly with a netsdk proto build.

### CODE REVIEW

File: src/FSharp.Compiler.Interactive.Settings/FSharp.Compiler.Interactive.Settings.fsproj [hunk 1]
```diff
@@ -11,8 +11,11 @@
   </PropertyGroup>
 
   <!--  The FSharp.Compiler.Interactive.Settings dll provides a referencable public interface to tool builders  -->
-  <PropertyGroup Condition="'$(Configuration)' != 'Proto'">
-      <CompressMetadata Condition="'$(CompressAllMetadata)' != 'true'">false</CompressMetadata>
+  <PropertyGroup>
+    <NoOptimizationData>true</NoOptimizationData>
+    <NoInterfaceData>false</NoInterfaceData>
+    <CompressMetadata>true</CompressMetadata>
+    <OtherFlags Condition="'$(Configuration)' != 'Proto'">$(OtherFlags) --realsig+</OtherFlags>
```
Comment (vzarytovskii) [inline src/FSharp.Compiler.Interactive.Settings/FSharp.Compiler.Interactive.Settings.fsproj:18]: Same here

File: src/fsc/fsc.targets [hunk 2]
```diff
@@ -13,10 +13,11 @@
...
   <!--  The fsc application does not provide a referencable public interface  -->
-  <PropertyGroup Condition="'$(Configuration)' != 'Proto'">
-      <NoOptimizationData>true</NoOptimizationData>
-      <NoInterfaceData>true</NoInterfaceData>
-      <CompressMetadata Condition="'$(CompressAllMetadata)' != 'true'">false</CompressMetadata>
+  <PropertyGroup>
+    <NoOptimizationData>true</NoOptimizationData>
+    <NoInterfaceData>true</NoInterfaceData>
+    <CompressMetadata>true</CompressMetadata>
+    <OtherFlags Condition="'$(Configuration)' != 'Proto'">$(OtherFlags) --realsig+</OtherFlags>
```
Comment (vzarytovskii) [inline src/fsc/fsc.targets:20]: Same

File: src/fsi/fsi.targets [hunk 3]
```diff
@@ -18,10 +18,11 @@
...
   <!--  The fsi application does not provide a referencable public interface  -->
-  <PropertyGroup Condition="'$(Configuration)' != 'Proto'">
-      <NoOptimizationData>true</NoOptimizationData>
-      <NoInterfaceData>true</NoInterfaceData>
-      <CompressMetadata Condition="'$(CompressAllMetadata)' != 'true'">false</CompressMetadata>
+  <PropertyGroup>
+    <NoOptimizationData>true</NoOptimizationData>
+    <NoInterfaceData>true</NoInterfaceData>
+    <CompressMetadata>true</CompressMetadata>
+    <OtherFlags Condition="'$(Configuration)' != 'Proto'">$(OtherFlags) --realsig+</OtherFlags>
```
Comment (vzarytovskii) [inline src/fsi/fsi.targets:25]: Same

File: tests/Directory.Build.props [hunk 4]
```diff
@@ -7,9 +7,16 @@
...
+  <!-- Do it the same way .NET Interactive does it, when skipping arcade, we should include test SDK explicitly -->
   <ItemGroup Condition="($(MSBuildProjectName.EndsWith('.Tests')) OR $(MSBuildProjectName.EndsWith('.ComponentTests')) OR $(MSBuildProjectName.EndsWith('.UnitTests'))) AND '$(DISABLE_ARCADE)' == 'true'">
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="$(MicrosoftNETTestSdkVersion)" />
   </ItemGroup>
 
+  <PropertyGroup>
+    <NoOptimizationData>false</NoOptimizationData>
+    <NoInterfaceData>false</NoInterfaceData>
+    <CompressMetadata>true</CompressMetadata>
+    <OtherFlags>$(OtherFlags) --realsig+</OtherFlags>
```
Comment (vzarytovskii) [inline tests/Directory.Build.props:19]: Same

File: vsintegration/Templates.Directory.Build.props [hunk 5]
```diff
@@ -31,4 +31,11 @@
     <CopyOutputSymbolsToOutputDirectory>false</CopyOutputSymbolsToOutputDirectory>
   </PropertyGroup>
 
+  <PropertyGroup>
+    <NoOptimizationData>true</NoOptimizationData>
+    <NoInterfaceData>false</NoInterfaceData>
+    <CompressMetadata>false</CompressMetadata>
+    <OtherFlags>$(OtherFlags) --realsig+</OtherFlags>
```
Comment (vzarytovskii) [inline vsintegration/Templates.Directory.Build.props:38]: Same

**Review by vzarytovskii — CHANGES_REQUESTED** [review]

File: vsintegration/Templates.Directory.Build.props (hunk 5 above)
Comment (KevinRansom) [reply vsintegration/Templates.Directory.Build.props:38]: @vzarytovskii Right at the moment, I am trying out some stuff, none of it is even close to final.

File: buildtools/AssemblyCheck/AssemblyCheck.fsproj [hunk 6]
```diff
@@ -7,10 +7,16 @@
     <UseAppHost Condition="'$(DotNetBuildFromSource)' == 'true'">false</UseAppHost>
   </PropertyGroup>
 
+  <PropertyGroup>
+    <PublishReadyToRun>true</PublishReadyToRun>
+    <RuntimeIdentifiers Condition="'$(OS)' == 'Unix'">linux-x64</RuntimeIdentifiers>
```
Comment (vzarytovskii) [inline buildtools/AssemblyCheck/AssemblyCheck.fsproj:12]: This will need Mac as well, right?

File: buildtools/fslex/fslex.fsproj [hunk 7]
```diff
@@ -8,6 +8,12 @@
     <RollForward Condition="'$(BUILDING_USING_DOTNET)' == 'true'">LatestMajor</RollForward>
   </PropertyGroup>
 
+  <PropertyGroup>
+    <PublishReadyToRun>true</PublishReadyToRun>
+    <RuntimeIdentifiers Condition="'$(OS)' == 'Unix'">linux-x64</RuntimeIdentifiers>
```
Comment (vzarytovskii) [inline buildtools/fslex/fslex.fsproj:13]: Same here for Mac

File: buildtools/fsyacc/fsyacc.fsproj [hunk 8]
```diff
@@ -8,6 +8,12 @@
     <RollForward Condition="'$(BUILDING_USING_DOTNET)' == 'true'">LatestMajor</RollForward>
   </PropertyGroup>
 
+  <PropertyGroup>
+    <PublishReadyToRun>true</PublishReadyToRun>
+    <RuntimeIdentifiers Condition="'$(OS)' == 'Unix'">linux-x64</RuntimeIdentifiers>
```
Comment (vzarytovskii) [inline buildtools/fsyacc/fsyacc.fsproj:13]: And here

**Review by vzarytovskii — CHANGES_REQUESTED** [review]

File: buildtools/AssemblyCheck/AssemblyCheck.fsproj (hunk 6 above)
Comment (KevinRansom) [reply buildtools/AssemblyCheck/AssemblyCheck.fsproj:12]: Hmm, that is a super great question.  Good eyes mate.

It works but I imagine on an M1 it runs emulated rather than arm native.
I think I should expand the palette of options trying to make it native as much as possible.

File: buildtools/fslex/fslex.fsproj (hunk 7 above)
Comment (KevinRansom) [reply buildtools/fslex/fslex.fsproj:13]: see above

File: buildtools/fsyacc/fsyacc.fsproj (hunk 8 above)
Comment (KevinRansom) [reply buildtools/fsyacc/fsyacc.fsproj:13]: see previous

File: buildtools/AssemblyCheck/AssemblyCheck.fsproj (hunk 6 above)
Comment (KevinRansom) [reply buildtools/AssemblyCheck/AssemblyCheck.fsproj:12]: @baronfel just pointed me at the  UseCurrentRuntimeIdentifier property which eliminates all thought and just does 'the right thing (TM)'.

I am a happy bunny

File: Build.cmd [hunk 9]
```diff
@@ -1,2 +1,2 @@
 @echo off
-powershell -ExecutionPolicy ByPass -NoProfile -command "& """%~dp0eng\build.ps1""" -build -restore %*"
\ No newline at end of file
+powershell -ExecutionPolicy ByPass -NoProfile -command "Measure-Command { & """%~dp0eng\build.ps1""" -build -restore %* } | Select-Object TotalMinutes, TotalSeconds"
```
Comment (T-Gro) [inline Build.cmd:2]: Are we keeping this change?
( I do not mind, I have similar wrappers in local profile anyway...)

File: src/FSharp.Build/Microsoft.FSharp.Targets [hunk 10]
```diff
@@ -365,6 +365,7 @@ this file.
               PreferredUILang="$(PreferredUILang)"
               ProvideCommandLineArgs="$(ProvideCommandLineArgs)"
               PublicSign="$(PublicSign)"
+              RealSig="$(RealSig)"
```
Comment (T-Gro) [inline src/FSharp.Build/Microsoft.FSharp.Targets:368]: Outside of proto and plain dotnet build, we should now be fine with using `<RealSig>true</>` in this repo, or?

**Review by T-Gro — APPROVED** [review]

File: Build.cmd (hunk 9 above)
Comment (KevinRansom) [reply Build.cmd:2]: We most certainly are. it's good :-)

File: buildtools/fsyacc/fsyacc.fsproj [hunk 11]
```diff
@@ -8,6 +8,11 @@
     <RollForward Condition="'$(BUILDING_USING_DOTNET)' == 'true'">LatestMajor</RollForward>
   </PropertyGroup>
 
+  <PropertyGroup>
+    <PublishReadyToRun>true</PublishReadyToRun>
```
[the remark this answers is not in the document]
Comment (KevinRansom) [reply buildtools/fsyacc/fsyacc.fsproj:12]: My preference is for the project to specify this stuff rather than directory.props and targets, it makes it clearer at least to me.  But I can move it if you really think it's better that way.

File: src/FSharp.Build/Microsoft.FSharp.Targets (hunk 10 above)
Comment (KevinRansom) [reply src/FSharp.Build/Microsoft.FSharp.Targets:368]: Sigh!!!, vlad and some folk still like using building_using_dotnet, that build task doesn't have realsig yet.

File: src/Compiler/FSharp.Compiler.Service.fsproj [hunk 12]
```diff
@@ -25,12 +25,19 @@
...
   <PropertyGroup Condition="'$(Configuration)' != 'Proto'">
-    <CompressMetadata Condition="'$(CompressAllMetadata)' != 'true'">false</CompressMetadata>
+    <NoOptimizationData>false</NoOptimizationData>
+    <NoInterfaceData>false</NoInterfaceData>
+    <CompressMetadata>false</CompressMetadata>
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Configuration)' != 'Proto' and '$(BUILDING_USING_DOTNET)' != 'true'">
+    <OtherFlags Condition="'$(BUILDING_USING_DOTNET)' != 'true' and '$(TestingLegacyInternalSignature)' != 'true'">$(OtherFlags) --realsig-</OtherFlags>
+    <OtherFlags Condition="'$(BUILDING_USING_DOTNET)' != 'true' and '$(TestingLegacyInternalSignature)' == 'true'">$(OtherFlags) --realsig+</OtherFlags>
```
Comment (T-Gro) [inline src/Compiler/FSharp.Compiler.Service.fsproj:40]: Also, isn't the logical condition here inverted?
Shouldn't `TestingLegacyInternalSignature` being true mean we want to test the old legacy signature, and therefore `---realsig-` effectively?
Just from the sounding of the phrase TestingLegacyInternalSignature...

File: src/fsc/fsc.targets [hunk 13]
```diff
@@ -4,19 +4,18 @@
...
   <!--  The fsc application does not provide a referencable public interface  -->
-  <PropertyGroup Condition="'$(Configuration)' != 'Proto'">
-      <NoOptimizationData>true</NoOptimizationData>
-      <NoInterfaceData>true</NoInterfaceData>
-      <CompressMetadata Condition="'$(CompressAllMetadata)' != 'true'">false</CompressMetadata>
+  <PropertyGroup>
+    <NoOptimizationData>true</NoOptimizationData>
+    <NoInterfaceData>true</NoInterfaceData>
+    <CompressMetadata>true</CompressMetadata>
+    <OtherFlags Condition="'$(BUILDING_USING_DOTNET)' != 'true' and '$(BUILDING_USING_DOTNET)' != 'true' and '$(Configuration)' != 'Proto'">$(OtherFlags) --realsig+</OtherFlags>
```
Comment (T-Gro) [inline src/fsc/fsc.targets:18]: The condition is repeated here.

```suggestion
    <RealSig Condition=" '$(BUILDING_USING_DOTNET)' != 'true' and '$(Configuration)' != 'Proto'">true</RealSig >
```

File: src/Compiler/FSharp.Compiler.Service.fsproj [hunk 14]
```diff
@@ -25,12 +25,19 @@
...
   <!--  The FSharp.Compiler.Service dll provides a referencable public interface for tool builders -->
   <PropertyGroup Condition="'$(Configuration)' != 'Proto'">
-    <CompressMetadata Condition="'$(CompressAllMetadata)' != 'true'">false</CompressMetadata>
+    <NoOptimizationData>false</NoOptimizationData>
+    <NoInterfaceData>false</NoInterfaceData>
+    <CompressMetadata>false</CompressMetadata>
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Configuration)' != 'Proto' and '$(BUILDING_USING_DOTNET)' != 'true'">
+    <OtherFlags Condition="'$(BUILDING_USING_DOTNET)' != 'true' and '$(TestingLegacyInternalSignature)' != 'true'">$(OtherFlags) --realsig-</OtherFlags>
```
[the remark this answers is not in the document]
Comment (KevinRansom) [reply src/Compiler/FSharp.Compiler.Service.fsproj:39]: We will start using the realsig build tag, once the rtm SDK and VS contains a buildtask that recognizes it.  Otherwise it's just too confusing to keep straight.
However, I can certainly remove the redundent '$(BUILDING_USING_DOTNET)' != 'true' tests.

File: src/fsc/fsc.targets (hunk 13 above)
Comment (KevinRansom) [reply src/fsc/fsc.targets:18]: This is a good suggestion, however, until the RTM build task supports the realsig build property, I would rather go the simpler route of using otherflags.

Comment (KevinRansom) [reply src/fsc/fsc.targets:18]: However, I can certainly remove the redundent '$(BUILDING_USING_DOTNET)' != 'true' tests.

File: src/Compiler/FSharp.Compiler.Service.fsproj (hunk 12 above)
Comment (KevinRansom) [reply src/Compiler/FSharp.Compiler.Service.fsproj:40]: The name and the flags do seem at variance don't they.

Comment (KevinRansom) [reply src/Compiler/FSharp.Compiler.Service.fsproj]: Okay I reworked the whole realsig+/- thing.  I think it makes much more sense now.

**Review by T-Gro — APPROVED** [review]

File: eng/Build.ps1 [hunk 15]
```diff
@@ -552,7 +552,7 @@ try {
     }
 
     $script:BuildMessage = "Failure building product"
-    if ($restore -or $build -or $rebuild -or $pack -or $sign -or $publish -and -not $skipBuild) {
+    if ($restore -or $build -or $rebuild -or $pack -or $sign -or $publish -and -not $skipBuild -and -not $sourceBuild) {
```
[the remark this answers is not in the document]
Comment (KevinRansom) [reply eng/Build.ps1:555]: We have the command line switch ... but yes I take the point.

File: eng/SourceBuildPrebuiltBaseline.xml [hunk 16]
```diff
@@ -16,6 +16,14 @@
     <UsagePattern IdentityGlob="System.Diagnostics.EventLog/7.0.0" />
     <UsagePattern IdentityGlob="System.Security.Cryptography.ProtectedData/7.0.0" />
 
+    <UsagePattern IdentityGlob="Microsoft.AspNetCore.App.Runtime.linux-x64/8.0.1" />
+    <UsagePattern IdentityGlob="Microsoft.AspNetCore.App.Runtime.win-x64/8.0.1" />
+    <UsagePattern IdentityGlob="Microsoft.NETCore.App.Crossgen2.linux-x64/8.0.1" />
+    <UsagePattern IdentityGlob="Microsoft.NETCore.App.Host.win-x64/8.0.1" />
+    <UsagePattern IdentityGlob="Microsoft.NETCore.App.Runtime.linux-x64/8.0.1" />
+    <UsagePattern IdentityGlob="Microsoft.NETCore.App.Runtime.win-x64/8.0.1" />
```
Comment (vzarytovskii) [inline eng/SourceBuildPrebuiltBaseline.xml:24]: Why are those needed now?

File: src/Compiler/FSharp.Compiler.Service.fsproj [hunk 17]
```diff
@@ -25,12 +25,14 @@
...
-    <Configurations>Debug;Release;ReleaseCompressed</Configurations>
+    <Configurations>Debug;Release</Configurations>
   </PropertyGroup>
 
   <!--  The FSharp.Compiler.Service dll provides a referencable public interface for tool builders -->
   <PropertyGroup Condition="'$(Configuration)' != 'Proto'">
-    <CompressMetadata Condition="'$(CompressAllMetadata)' != 'true'">false</CompressMetadata>
+    <NoOptimizationData>false</NoOptimizationData>
+    <NoInterfaceData>false</NoInterfaceData>
+    <CompressMetadata>false</CompressMetadata>
```
Comment (vzarytovskii) [inline src/Compiler/FSharp.Compiler.Service.fsproj:35]: Why are those needed now explicitly?

File: src/FSharp.Build/FSharp.Build.fsproj [hunk 18]
```diff
@@ -13,13 +13,14 @@
...
   </PropertyGroup>
 
   <!--  The FSharp.Build dll does not provide a referencable public interface although it's used for testing -->
-  <PropertyGroup Condition="'$(Configuration)' != 'Proto'">
-      <CompressMetadata Condition="'$(CompressAllMetadata)' != 'true'">false</CompressMetadata>
-      <NoOptimizationData>true</NoOptimizationData>
+  <PropertyGroup>
+    <NoOptimizationData>true</NoOptimizationData>
+    <NoInterfaceData>false</NoInterfaceData>
+    <CompressMetadata>true</CompressMetadata>
```
Comment (vzarytovskii) [inline src/FSharp.Build/FSharp.Build.fsproj:23]: Same question here

File: src/FSharp.Core/FSharp.Core.fsproj [hunk 19]
```diff
@@ -4,7 +4,8 @@
 
   <PropertyGroup>
     <OutputType>Library</OutputType>
-    <TargetFrameworks>netstandard2.1;netstandard2.0</TargetFrameworks>
+    <TargetFrameworks Condition="'$(Configuration)' == 'Proto'">netstandard2.0</TargetFrameworks>
```
Comment (vzarytovskii) [inline src/FSharp.Core/FSharp.Core.fsproj:7]: Should we build ns21 for proto?

**Review by vzarytovskii — APPROVED** [review]

File: src/FSharp.Core/FSharp.Core.fsproj (hunk 19 above)
Comment (KevinRansom) [reply src/FSharp.Core/FSharp.Core.fsproj:7]: @vzarytovskii No ... because the fsharp.build is loaded into msbuild, which is part of VS and can run on the desktop.

File: eng/SourceBuildPrebuiltBaseline.xml (hunk 16 above)
Comment (KevinRansom) [reply eng/SourceBuildPrebuiltBaseline.xml:24]: Because we now native compile proto and coreclr fsc and fsi rather than just build to IL.

File: src/Compiler/FSharp.Compiler.Service.fsproj (hunk 17 above)
Comment (KevinRansom) [reply src/Compiler/FSharp.Compiler.Service.fsproj:35]: Okay double negative, of course it has OptimizeData and InterfaceData so I understand that.

As for why it's needed here:  I am trying to be explicit about these settings, to make clear what setting is needed for each project.

File: src/FSharp.Build/FSharp.Build.fsproj (hunk 18 above)
Comment (KevinRansom) [reply src/FSharp.Build/FSharp.Build.fsproj:23]: I am trying to be explicit about these settings, to make clear what setting is needed for each project.  It is basically fixing a previously missing specification, rather than relying on unset values.

**Review by vzarytovskii — APPROVED** [review]

File: eng/SourceBuildPrebuiltBaseline.xml (hunk 16 above)
Comment (T-Gro) [reply eng/SourceBuildPrebuiltBaseline.xml:24]: Why is anything with `AspNetCore` in this list though?

**Review by MichaelSimons — APPROVED** [review]
